toolchain/semantics/semantics_handle_index.cpp

```diff
@@ -2,20 +2,65 @@
 // Exceptions. See /LICENSE for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-#include <utility>
-
 #include "toolchain/semantics/semantics_context.h"
+#include "toolchain/semantics/semantics_node.h"
+#include "toolchain/semantics/semantics_node_kind.h"
 
 namespace Carbon {
 
-auto SemanticsHandleIndexExpressionStart(SemanticsContext& context,
-                                         ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleIndexExpressionStart");
+auto SemanticsHandleIndexExpressionStart(SemanticsContext& /*context*/,
+                                         ParseTree::Node /*parse_node*/)
+    -> bool {
+  // Leave the expression on the stack for IndexExpression.
+  return true;
 }
 
 auto SemanticsHandleIndexExpression(SemanticsContext& context,
                                     ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleIndexExpression");
+  auto ind_node_id = context.node_stack().PopExpression();
+  auto name_node_id = context.node_stack().PopExpression();
+  auto name_node = context.semantics_ir().GetNode(name_node_id);
+
+  if (context.semantics_ir().GetNode(ind_node_id).kind() ==
+      SemanticsNodeKind::VarStorage) {
+    CARBON_DIAGNOSTIC(NondeterministicType, Error,
+                      "Type cannot be determined in compile time.");
+    context.emitter().Emit(parse_node, NondeterministicType);
+  } else if (name_node.kind() != SemanticsNodeKind::VarStorage) {
+    CARBON_DIAGNOSTIC(InvalidIndexExpression, Error,
+                      "Invalid index expression.");
+    context.emitter().Emit(parse_node, InvalidIndexExpression);
```

Comment:
Let's consider another case:

```
fn F() -> (i32,) { return (0,); }

fn Run() -> i32 {
  return F()[0];
}
```

You should probably be looking at the type of the node, not the kind.